add traefik-route endpoint

PietroPasotti opened the pull request:
## Feature

This PR adds an integration with the Traefik Route charm (cf. https://github.com/PietroPasotti/traefik-route-k8s-operator), enabling users to configure Traefik on a per-relation basis, overcoming the limitations of ingress-per-unit and ingress-per-app, making Traefik a lot more general.

The role of Traefik in this integration is very limited: it simply has to accept a yaml config file and write it to the folder where traefik will load it up. Therefore the changes to the charm codebase are rather limited.

## Testing instructions
Aside from running the utests, you can test this in combination with TraefikRoute and prometheus-k8s in a live deployment (itests pending).

juju deploy traefik-route-k8s --channel beta

## Release notes
- Added `traefik_route_k8s/v0/traefik_route.py` lib
- Added `traefik_route` integration in the Traefik charm.

simskij commented on src/charm.py:
```diff
@@ -157,7 +159,7 @@ def _on_traefik_pebble_ready(self, _: PebbleReadyEvent):
                 # We always start the Prometheus endpoint for simplicity
                 # TODO: Generate this file in the dynamic configuration folder when the
-                # metrics-endpoint relation is established?
+                #  metrics-endpoint relation is established?
```
```suggestion
                # metrics-endpoint relation is established?
```

PietroPasotti replied:
that was intentional. Convention is to indent multiline todo's and fixme's to make them count as one. It might be just a pycharm thing however.

PietroPasotti replied:
agreed for consistency with other TODOs/FIXMEs in the file

simskij commented on src/charm.py:
```diff
@@ -254,6 +256,14 @@ def _handle_ingress_request(self, event: RelationEvent):
+        """A traefik_route charm has published some ingress data."""
+        self._process_ingress_relation(event.relation)
+
+        # go to active if we're in maintenance
```
```suggestion
```

obvious from the code

PietroPasotti commented:
@jnsgruk I dropped #44 and replaced it with this one because rebasing gave me tons of conflicts, this was the easiest way to squash some commits and get a cleaner diff

jnsgruk reviewed (approved):
LGTM - thanks for the iteration on this.
